### PR TITLE
Change SkyModels and Datasets inheritance to MutableSequence

### PR DESCRIPTION
**Pull request by QRemy**
- change SkyModels and Datasets inheritance to MutableSequence in order to gain .append(), .extend(), .pop(), .remove() methods as discussed in #2678

### CODE REVIEW

File: gammapy/modeling/models/core.py [hunk 1]
```diff
@@ -182,22 +183,43 @@ def __str__(self):
 
     def __add__(self, other):
```
Comment (adonath) [inline gammapy/modeling/models/core.py]: I think the `__add__` method could call `self.extend()` and `self.append()` now.

File: gammapy/modeling/models/core.py [hunk 2]
```diff
@@ -182,22 +183,43 @@ def __str__(self):
...
-            return self._models[val]
-        elif isinstance(val, str):
+    def __getitem__(self, key):
+        return self._models[self._get_idx(key)]
+
+    def __delitem__(self, key):
+        del self._models[self._get_idx(key)]
+
+    def __setitem__(self, key, model):
+        if model.name in self.names:
```
Comment (adonath) [inline gammapy/modeling/models/core.py]: I think we check the mode instance here like `if isinstance(model, (SkyModel, SkyDiffuseCube))` and throw an error else.

File: gammapy/modeling/models/core.py [hunk 3]
```diff
@@ -182,22 +183,43 @@ def __str__(self):
...
+            raise (ValueError("Model names must be unique"))
+        self._models[self._get_idx(key)] = model
+
+    def insert(self, idx, model):
+        if model.name in self.names:
```
Comment (adonath) [inline gammapy/modeling/models/core.py]: Same here, please add an `isinstance()` check.

File: gammapy/modeling/datasets.py [hunk 4]
```diff
@@ -257,16 +257,32 @@ def info_table(self, cumulative=False):
...
-            return self._datasets[val]
-        elif isinstance(val, str):
+    def __getitem__(self, key):
+        return self._datasets[self._get_idx(key)]
+
+    def __delitem__(self, key):
+        del self._datasets[self._get_idx(key)]
+
+    def __setitem__(self, key, dataset):
+        if dataset.name in self.names:
```
Comment (adonath) [inline gammapy/modeling/datasets.py]: Maybe add an `isinstance(dataset, Dataset)` check here?

File: gammapy/modeling/datasets.py [hunk 5]
```diff
@@ -257,16 +257,32 @@ def info_table(self, cumulative=False):
...
+            raise (ValueError("Dataset names must be unique"))
+        self._datasets[self._get_idx(key)] = dataset
+
+    def insert(self, idx, dataset):
+        if dataset.name in self.names:
```
Comment (adonath) [inline gammapy/modeling/datasets.py]: Also add an `isinstance()` check here?

File: gammapy/modeling/models/core.py (hunk 1 above)
Comment (QRemy) [reply gammapy/modeling/models/core.py]: Not changed, as we want to return a new object rather than updating the first one

**Pull request merged**